dashboard/src/modules/components/SearchComponent/index.jsx

```diff
@@ -11,28 +10,24 @@ function SearchBox({
   setControllerName,
 }) {
   const [controllerValue, setControllerValue] = useState("");
-  let modifiedArray = [];
   const searchController = () => {
+    let modifiedArray = [];
     modifiedArray = dataArray.filter((data) => {
-      let formattedData = moment(data.metadata["dataset.created"]).format(
-        "YYYY/MM/DD"
-      );
       return (
         data.controller.includes(controllerValue) &&
-        Date.parse(formattedData) >= Date.parse(startDate) &&
-        Date.parse(formattedData) <= Date.parse(endDate)
+          new Date((data.metadata["dataset.created"]).split(":")[0])>= startDate &&
+          new Date((data.metadata["dataset.created"]).split(":")[0])<= new Date(endDate)
```

Comment:
I suspect that this code's operation is depending on Javascript implicitly converting between `Date`'s and strings.

As I said above, I think we should either be comparing them all as strings (which, given ISO and/or JSON formatting conventions, should work fine, at least once the time has been split out) or they should all be converted to `Date`'s.

Moreover, this code seems to duplicate code in `DatePicker`...I wonder if it shouldn't be refactored into a helper function (which we can keep with the date formatter function...).

Reply:
@dbutenhof , so I checked the DatePicker Patternfly code (which i used in my application), and when we select the from date from it,it is in the Javascript Date format ,while the to date is in a string format when selected from Date picker!! Now, i tried to convert the string date to a Date object,but it throws an error:- 
`val.split is not a function` in DatePicker.tsx module

So I am guessing that internally patternfly wants the to date in a string format only and hence if we want to compare dates in Date format ,than we have to implicitly convert the toDate to a Date object(which I have done here subsequently)

I am attaching the Link to the patternfly Datepicker doc so that you can check this out
https://www.patternfly.org/v4/components/date-picker/react-demos

Reply:
Thanks for the documentation link!

It looks like the `<DatePicker>`  `onChange` callback provides two parameters, a string and a `Date` (and, I'm hoping that they are both representations of the picked date...).  The `onFromChange()` function receives them both (but it seems to ignore the string); the analog for `toDate` somehow only uses one, which either magically turns out to be the `Date` or it's implicitly converting the string to a `Date` (I don't speak Javascript well enough to say which).

So, it seems like you should be able to get a `Date` object from the `DatePicker` and store that in the React store (for each of the `fromDate` and the `toDate` values), and we don't need to mess about with strings at all (other than converting the JSON-format dates that come from the server into Javascript `Date` objects).

Reply:
Let me expand....

Here's some of the code from the Patternfly web page:

```
  const onFromChange = (_str, date) => {
    setFrom(new Date(date));
    if (isValidDate(date)) {
      date.setDate(date.getDate() + 1);
      setTo(yyyyMMddFormat(date));
    }
    else {
      setTo('');
    }
  };
```
To my eye, this sets the `fromDate` with a `Date` and then sets the `toDate` with the result of `yyyyMMddFormat()` which I expect is a **_string_**.  So, the two values end up with inconsistent types.  Personally, I would object to this....

In the `DatePicker` for the `toDate`, the Patternfly code is
```
          onChange={date => setTo(date)}
```
but, I'm not sure what the type of `date` is in this context.  If it is the first value specified by the callback, then it is most probably a string, and so the code is setting `toDate` to a string, which is consistent with what the `fromDate` callback does, but, inconsistent with the type of `fromDate`.  However, if `date` is in fact bound to the second parameter of the callback (e.g., if Javascript is doing a by-name binding here, which I suspect that it might be), then in fact `date` is probably a `Date` and so `toDate` gets set to a `Date`, which is would be consistent with `fromDate` but inconsistent with the setting that it receives when the other callback is invoked.

So, I would fix the Patternfly code, and then see how things go.  😁

Reply:
`onChange={date => setTo(date)}`
Yes,the date here is a string ,and as you mentioned ,it is inconsistent with the type of fromDate.Even I object to this,since it is a unnecessary inconsistency which I guess can be easily removed by modifying few things.I was really intrigued why it was written in such a way in Patternfly and tried to reason out some stuff but it didn’t made much sense.

Okk,I will fix the patternfly code and see if things work or not.

Reply:
I finally found the [doc for `DatePicker`](https://www.patternfly.org/v4/components/date-picker/), and it says the signature for the `onChange` callback is `(value: string, date?: Date) => void`.

I found the [caller of the `onChange()` callback](https://github.com/patternfly/patternfly-react/blob/cc80f358e790331d92df19b49df25488b095265b/packages/react-core/src/components/DatePicker/DatePicker.tsx#L118):
```
  const onTextInput = (value: string) => {
    setValue(value);
    setErrorText('');
    const newValueDate = dateParse(value);
    setValueDate(newValueDate);
    if (isValidDate(newValueDate)) {
      onChange(value, new Date(newValueDate));
    } else {
      onChange(value);
    }
  };
```
So, as I surmised, if the date value is invalid, the callback receives only the string; otherwise, it gets both the string and the corresponding `Date` object.

So, if you can fix the two callback functions so that they capture the _date_ parameter value, then we don't really have to worry about strings.  However, your code should probably check that the value is not null or undefined, just to cover the case where the user manually entered something which turned out to be a bad value.

Reply:
As you and I discussed yesterday, @shubham-html-css-js, contrary to my previous comment, we probably need to operate on the string value and mostly ignore the `Date` object:  the `Date` object will represent (in UTC) a _local time_, whereas we want the user to select UTC time; so, we need to take the string (which represents what the user actually entered or selected) and produce a `Date` object corresponding to it treated as UTC time.  (Also, this will likely allow us to better accommodate the case when the `Date` object is `undefined` because the string doesn't represent a valid time.)

Reply:
Can we consider handling all these data changes in a follow-up PR?

Reply:
[Not really](https://github.com/distributed-system-analysis/pbench/pull/2733#discussion_r862843082).